notifier/notify_windows.go

```diff
@@ -1,16 +1,290 @@
+		for _, actions := range n.AvailableActions {
+			if actions.Text == data["button"] {
+				n.SelectAction(actions.ID)
+			}
+		}
+	default:
+		log.Debugf("SnoreToast-Action %s not implemented yet", data["action"])
+	}
+}
+
+//Snoretoast helper-Functions:
+
+func snoreToastStrToMap(in string) map[string]string {
+	ret := make(map[string]string)
+
+	for _, el := range strings.Split(in, ";") { //TODO: make sure, there is no ; in any value!
+		if el == "" {
+			continue
+		}
+
+		kv := strings.Split(el, "=") //TODO: make sure, there is no = in any value!
+		if len(kv) != 2 {
+			log.Warning("SnoreToast-Callback: " + el + " cannot be parsed in Key=Value")
+			continue
+		}
+
+		ret[kv[0]] = kv[1]
+	}
+
+	return ret
+}
+
+func buildSnoreToastButtonArgument(a []*Action) string {
+	temp := make([]string, 0, len(a))
+	for _, elem := range a {
+		if elem.Text != "" {
+			temp = append(temp, elem.Text)
+		}
+	}
+	return strings.Join(temp, ";")
+}
+
+func snoreToastRunCmdAsync(cmd *exec.Cmd) {
+	go func() {
+		exit, err := execCmd(cmd)
+
+		switch exit {
+		case 0: //do nothing
+		case 1: //do nothing
+		case 2: //do nothing
+		case 3: //do nothing
+		case 4: //do nothing
+		case 5: //do nothing
+		default:
+			log.Errorf("Error while running %+v: %s", cmd.Args, err)
+		}
+	}()
+}
+
+//Generel helper-Functions:
+
+func execCmd(cmd *exec.Cmd) (exitCode int, err error) {
+	log.Debugf("Run snoretoast:%+v", cmd.Args)
+
+	err = cmd.Run()
+
+	//Used: https://stackoverflow.com/questions/10385551/get-exit-code-go
+	if err != nil {
+		exitError, ok := err.(*exec.ExitError)
+		if ok {
+			ws := exitError.Sys().(syscall.WaitStatus)
+			exitCode = ws.ExitStatus()
+		} else {
+			// This will happen (in OSX) if `name` is not available in $PATH,
+			// in this situation, exit code could not be get, and stderr will be
+			// empty string very likely, so we use the default fail code, and format err
+			// to string and set to stderr
+			log.Warningf("Could not get exit code for failed program: %v", cmd.Args)
+			exitCode = -2 //We want to have something that sticks out somehow
+		}
+	} else {
+		// success, exitCode should be 0 if go is ok
+		ws := cmd.ProcessState.Sys().(syscall.WaitStatus)
+		exitCode = ws.ExitStatus()
+	}
+
+	return
+}
+
+func readWstrConnToStr(conn net.Conn) string {
+	defer conn.Close()
+
+	var bufferslice []byte
+
+readloop:
+	for {
+		buffer := make([]byte, 512)
+
+		n, err := conn.Read(buffer)
+		switch err {
+		case nil:
+			//do nothing
+		case io.EOF:
+			break readloop
+		default:
+			log.Warningf("%s", err)
+			return ""
+		}
+		bufferslice = append(bufferslice, buffer[:n]...)
+	}
+
+	dec := unicode.UTF16(unicode.LittleEndian, unicode.UseBOM).NewDecoder()
+	out, err := dec.Bytes(bufferslice)
+	if err != nil {
+		log.Warningf("Error while converting wstr to str: %s", err)
+		return ""
+	}
+
+	return string(out)
+}
+
+func getPath(what string) (string, error) {
+	// build path to app
+	if dataDir == "" {
+		log.Errorf("dataDir is empty!!!")
```

Comment:
keep your voice down. ;)

Please provide more context for this error message. If this really is a serious error, it should be clear what exactly is the problem when reading the logs.

Actually, this error should probably returned instead of logging the error.

If this is just a debugging message you forgot, please remove it. In future, please use `log.Trace*` for these kinds of (temporary) log messages.